c++/src/kj/compat/tls.c++

```diff
@@ -405,31 +416,108 @@ private:
 // =======================================================================================
 // Implementations of ConnectionReceiver, NetworkAddress, and Network as wrappers adding TLS.
 
-class TlsConnectionReceiver final: public kj::ConnectionReceiver {
+template <typename T>
+class InputQueue {
+  // InputQueue is an async FIFO queue that has many sources but only one sink.
+  // TODO(pr) Should I find a header home for this somewhere?
+
 public:
-  TlsConnectionReceiver(TlsContext& tls, kj::Own<kj::ConnectionReceiver> inner)
-      : tls(tls), inner(kj::mv(inner)) {}
+  void pushRaw(T&& v) {
+    // Push an existing value onto the queue.
 
-  Promise<Own<AsyncIoStream>> accept() override {
-    return inner->accept().then([this](kj::Own<AsyncIoStream> stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return accept();
-      });
+    if (auto localFulfiller = kj::mv(fulfiller)) {
+      // Fast path: We have a consumer waiting.
+      localFulfiller->fulfill(kj::fwd<T>(v));
+      return;
+    }
+
+    // Slow path: We don't have a consumer, add it to the queue.
+    auto node = kj::heap<Node>(kj::fwd<T>(v));
+    if (head) {
+      Node* currentNode = head.get();
+      while (currentNode->next) {
+        // Run through the nodes until we don't have a next.
+        currentNode = currentNode->next.get();
+      }
+
+      currentNode->next = kj::mv(node);
+    } else {
+      head = kj::mv(node);
+    }
+  }
+
+  Promise<void> push(Promise<T>&& p) {
+    // Eventually push a value onto the queue.
+
+    return kj::fwd<Promise<T>>(p).then([this](T&& v) {
```

Comment:
The `kj::fwd<Promise<T>>(p)` doesn't seem to be accomplishing anything here, vs. just `p`.

Reply:
Mind if I leave it as `kj::mv`? I can see a world where we can optimize promises based on cv qualifiers.

Reply:
`.then()` (and most other methods of `Promise`) already moves the promise away implicitly. That's not something we could change now without breaking the world. So using `kj::mv` is redundant.